src/TeaPie/Logging/RequestsLoggingHandler.cs

```diff
@@ -0,0 +1,165 @@
+﻿using System.Net.Http.Headers;
+using TeaPie.Http;
+using TeaPie.Http.Auth;
+using Microsoft.Extensions.Logging;
+
+namespace TeaPie.Logging;
+
+internal class RequestsLoggingHandler(IAuthProviderAccessor authProviderAccessor, ILoggerFactory loggerFactory) : DelegatingHandler
+{
+    private readonly IAuthProviderAccessor _authProviderAccessor = authProviderAccessor;
+    private readonly ILogger _logger = loggerFactory.CreateLogger("HttpRequests");
+    private static readonly HttpRequestOptionsKey<RequestExecutionContext> _contextKey = new("__TeaPie_Context__");
+    public static readonly HttpRequestOptionsKey<RequestLogFileEntry> LogEntryKey = new("__TeaPie_LogEntry__");
+
+    protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+    {
+        if (!request.Options.TryGetValue(_contextKey, out var requestContext) || requestContext is null)
+        {
+            return await base.SendAsync(request, cancellationToken);
+        }
+
+        var logEntry = await GetOrCreateLogEntryAsync(request, requestContext);
+        var attemptStartTime = DateTime.UtcNow;
+        HttpResponseMessage? response = null;
+        Exception? exception = null;
+
+        try
+        {
+            response = await base.SendAsync(request, cancellationToken);
+            return response;
+        }
+        catch (Exception ex)
+        {
+            exception = ex;
+            throw;
+        }
+        finally
+        {
+            await RecordAttemptAsync(logEntry, response, exception, attemptStartTime);
+            await LogCompletedRequestAsync(request, response);
+        }
+    }
+
+    private async Task<RequestLogFileEntry> GetOrCreateLogEntryAsync(HttpRequestMessage request, RequestExecutionContext requestContext)
+    {
+        if (request.Options.TryGetValue(LogEntryKey, out var existingEntry) && existingEntry != null)
```

Comment:
nie je toto celé ešte pozostatok kvôli tomu, že si chcela špecificky logovať tie retries?
Nejak mám pocit, že tým ako si to teraz pekne učesala do tohto delegate handlera a tým, že som povedal že retryis netreba špecificky spracovávať tak toto prenášanie si vecí cez Option je už zbytočné. 

Ale fakt neviem, môžem sa mýliť.

Reply:
🤔 alebo je to kvôli tomu aby si mala teda to unikátne IDčko? To mi došlo až teraz, je to tak?